drivers/gpio/gpio_mcp230xx.c

```diff
@@ -0,0 +1,377 @@
+		(struct mcp230xx_drv_data *)dev->data;
+	uint16_t buf;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	ret = read_port_regs(dev, REG_GPIO, &buf);
+	if (ret != 0) {
+		goto done;
+	}
+
+	*value = buf;
+
+done:
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp230xx_port_set_masked_raw(const struct device *dev,
+					uint32_t mask, uint32_t value)
+{
+	struct mcp230xx_drv_data *drv_data =
+		(struct mcp230xx_drv_data *)dev->data;
+	uint16_t buf;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	buf = drv_data->reg_cache.gpio;
+	buf = (buf & ~mask) | (mask & value);
+
+	ret = write_port_regs(dev, REG_GPIO, buf);
+	if (ret == 0) {
+		drv_data->reg_cache.gpio = buf;
+	}
+
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp230xx_port_set_bits_raw(const struct device *dev, uint32_t mask)
+{
+	return mcp230xx_port_set_masked_raw(dev, mask, mask);
+}
+
+static int mcp230xx_port_clear_bits_raw(const struct device *dev,
+					uint32_t mask)
+{
+	return -mcp230xx_port_set_masked_raw(dev, mask, 0);
+}
+
+static int mcp230xx_port_toggle_bits(const struct device *dev, uint32_t mask)
+{
+	struct mcp230xx_drv_data *drv_data =
+		(struct mcp230xx_drv_data *)dev->data;
+	uint16_t buf;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	buf = drv_data->reg_cache.gpio;
+	buf ^= mask;
+
+	ret = write_port_regs(dev, REG_GPIO, buf);
+	if (ret == 0) {
+		drv_data->reg_cache.gpio = buf;
+	}
+
+	k_sem_give(&drv_data->lock);
+
+	return ret;
+}
+
+static int mcp230xx_pin_interrupt_configure(const struct device *dev,
+					    gpio_pin_t pin,
+					    enum gpio_int_mode mode,
+					    enum gpio_int_trig trig)
+{
+	return -ENOTSUP;
+}
+
+static const struct gpio_driver_api api_table = {
+	.pin_configure = mcp230xx_pin_cfg,
+	.port_get_raw = mcp230xx_port_get_raw,
+	.port_set_masked_raw = mcp230xx_port_set_masked_raw,
+	.port_set_bits_raw = mcp230xx_port_set_bits_raw,
+	.port_clear_bits_raw = mcp230xx_port_clear_bits_raw,
+	.port_toggle_bits = mcp230xx_port_toggle_bits,
+	.pin_interrupt_configure = mcp230xx_pin_interrupt_configure,
+};
+
+/**
+ * @brief Initialization function of MCP230XX
+ *
+ * @param dev Device struct.
+ * @return 0 if successful. Otherwise <0 is returned.
+ */
+static int mcp230xx_init(const struct device *dev)
+{
+	const struct mcp230xx_config *config = dev->config;
+	struct mcp230xx_drv_data *drv_data =
+		(struct mcp230xx_drv_data *)dev->data;
+
+	if (config->ngpios != 8 && config->ngpios != 16) {
+		LOG_ERR("Invalid value ngpios=%u. Expected 8 or 16!", config->ngpios);
+		return -EINVAL;
+	}
+
+	drv_data->i2c = device_get_binding((char *)config->i2c_dev_name);
+	if (!drv_data->i2c) {
+		LOG_ERR("Unable to get SPI device");
+		return -ENODEV;
+	}
+
+	drv_data->i2c_cfg = I2C_SPEED_SET(I2C_SPEED_STANDARD) | I2C_MODE_MASTER;
```

Comment:
The speed setting should come from DTS.

Reply:
Which setting should be used? Maybe `clock-frequency` from i2c-controller works? The driver also works if the call to i2c_configure is removed so maybe this setting should be removed anyways.

Reply:
It's indeed `clock-frequency` property. The I2C API doesn't state it clearly however it seems all the I2C drivers perform a call to their local `i2c_configure` function as part of the driver initialization. In this case it's best to remove the call to `i2c_configure` from this driver, just as you did.